File: MedusaTaskProxy.m
//
//  MedusaTaskProxy.m
//
//  Created by Tomasz Kukielka on 9/20/20.
//  Copyright © 2020 Tomasz Kukielka. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "MedusaTaskProxy.h"

//#define TRACE_PROXY 1

@interface MedusaTaskProxy()
	@property(nonatomic, strong) dispatch_block_t taskBlock;
@end

@implementation MedusaTaskProxy

- (id)initWithTask:(dispatch_block_t)task
{
	self = [super init];
	if(self != nil)
	{
		_taskBlock = task;
		_executed = false;
	}

	return self;
}


- (void)executeTask
{
#if TRACE_PROXY
	printf("executing proxy = %p\n", (__bridge void *)self);
#endif

	// there must be no mistake in algorithm
	// no matter what, the task in graph can only be executed once
	assert(!_executed);
	
	@autoreleasepool
	{
		//execute the actual requested task now
		_taskBlock();
		
		_executed = true;
		
		// free some memory
		// in single-threaded recurive algorithm it does not have a big perf penalty
		_taskBlock = NULL;
	}
}

#if TRACE_PROXY
-(void)dealloc
{
	 printf("dealloc proxy = %p\n", (__bridge void *)self);
}
#endif


#if ENABLE_DEBUG_DUMP

- (void)dumpDescription
{
	printf("MedusaTaskProxy self=%p\n", (__bridge void *)self);
	
	printf("  inputs:\n");
	for(NSUInteger i = 0; i < _inputCount; i++)
	{
		printf("    ");
		DumpBranchForNode(_inputs[i]);
	}

	printf("  outputs:\n");
	for(NSUInteger i = 0; i < _outputCount; i++)
	{
		printf("    ");
		DumpBranchForNode(_outputs[i]);
	}
}
#endif //ENABLE_DEBUG_DUMP

@end //@implementation MedusaTaskProxy

File: medusa-objc/MedusaTaskProxy.cpp
#include "MedusaTaskProxy.h"
#include <cassert>
#include <cstdlib>

//#define TRACE_PROXY 1

MedusaTaskProxy::MedusaTaskProxy(std::function<void()> task)
	: taskBlock(std::move(task))
{
}

MedusaTaskProxy::~MedusaTaskProxy()
{
#if TRACE_PROXY
	printf("dealloc MedusaTaskProxy = %p\n", this);
#endif
	free(inputs);
	free(outputs);
}

void MedusaTaskProxy::executeTask()
{
#if TRACE_PROXY
	printf("executing MedusaTaskProxy = %p\n", this);
#endif
	assert(!executed);

	taskBlock();
	executed = true;

	// Single-threaded path: freeing here is cheap (no cross-thread penalty).
	taskBlock = nullptr;
}

#if ENABLE_DEBUG_DUMP
void MedusaTaskProxy::dumpDescription() const
{
	printf("MedusaTaskProxy self=%p\n", this);

	printf("  inputs:\n");
	for(size_t i = 0; i < inputCount; i++)
	{
		printf("    ");
		DumpBranchForNode(inputs[i]);
	}

	printf("  outputs:\n");
	for(size_t i = 0; i < outputCount; i++)
	{
		printf("    ");
		DumpBranchForNode(outputs[i]);
	}
}
#endif
